<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>&lt;no title&gt; &mdash; AIPyS  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> AIPyS
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="install/index.html">Installing AIPyS</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="API.html">AIPyS modules</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">AIPyS</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">&lt;no title&gt;</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/07-hit-enrichment-analysis.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<p>Our platform allows for screening rates of up to one million cells per run, with a minimum rate of 100,000 cells. To optimize this process, an enrichment strategy has been developed.</p>
<p>In a CRISPR perturbation screen, the Geometric Distribution is the most commonly used sgRNA sequencing distribution, with an emphasis on evenly distributes sgRNAs across a gene or set of genes. However, the Negative Binomial Distribution is also used and provides more accurate measurement of perturbation by emphasizing the number of times a gene target is hit.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">Simulate</span></code> module simulates AIPS by definingthe  the sgRNA pool for a given gene using the <code class="docutils literal notranslate"><span class="pre">lookupString</span></code> parameter. It uses the published crispri sgRNA data (Horlbeck et al) and a Negative Binomial distribution with parameters n and p to determine the number of effective (successful) sgRNA in the pool , as shown in the Python library <a class="reference external" href="https://www.pymc.io/welcome.html">pymc</a>. The argument <code class="docutils literal notranslate"><span class="pre">tpRatio</span></code> indicates the number of effective target sgRNA in the pool. It is known that approximately 0.2 to 0.4 of the sgRNA targeting a particular gene are effective, according to the paper by Daley et al.</p>
<ul class="simple">
<li><p>Max A Horlbeck, Luke A Gilbert, Jacqueline E Villalta, Britt Adamson, Ryan A Pak, Yuwen Chen, Alexander P Fields, Chong Yon Park, Jacob E Corn, Martin Kampmann, Jonathan S Weissman (2016) Compact and highly active next-generation libraries for CRISPR-mediated gene repression and activation eLife 5:e19760 https://doi.org/10.7554/eLife.19760</p></li>
<li><p>Daley, T., Lin, Z., Lin, X. et al. CRISPhieRmix: a hierarchical mixture model for CRISPR pooled screens. Genome Biol 19, 159 (2018). https://doi.org/10.1186/s13059-018-1538-6</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">AIPyS</span> <span class="kn">import</span> <span class="n">AIPS_simulate</span> <span class="k">as</span> <span class="n">sim</span>
<span class="c1"># filtered list of sgRNA data from the reference library for the sub library h4</span>
<span class="n">dfH4</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;dfH4.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The command <code class="docutils literal notranslate"><span class="pre">simulation</span></code> defines the false positives limits and the number of observations per acquisition, which is drawn from a normal distribution.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">orig</span><span class="p">,</span> <span class="n">Q1</span><span class="p">,</span> <span class="n">Q2</span> <span class="o">=</span> <span class="n">sim</span><span class="o">.</span><span class="n">Simulate</span><span class="p">(</span><span class="n">df</span> <span class="o">=</span> <span class="n">dfH4</span> <span class="p">,</span><span class="n">lookupString</span> <span class="o">=</span>  <span class="s1">&#39;PEX&#39;</span> <span class="p">,</span><span class="n">tpRatio</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span> <span class="mi">10</span> <span class="p">,</span> <span class="n">p</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">)</span><span class="o">.</span><span class="n">simulation</span><span class="p">(</span>
                            <span class="n">FalseLimits</span>  <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.5</span><span class="p">),</span> <span class="n">ObservationNum</span> <span class="o">=</span> <span class="p">(</span><span class="mi">70</span><span class="p">,</span><span class="mi">20</span><span class="p">))</span>
</pre></div>
</div>
<p>This function returns three data frames: <code class="docutils literal notranslate"><span class="pre">orig</span></code> contains the original read counts, <code class="docutils literal notranslate"><span class="pre">Q1</span></code> contains the read counts of the sgRNAs not selected during acquisition, and <code class="docutils literal notranslate"><span class="pre">Q2</span></code> contains the read counts of the selected hits the “activated” sample.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;activeSg&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">indexPexActiveArray</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">indexTarget</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">tpRatio</span><span class="p">)]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">indexPexActiveArray</span><span class="p">,</span> <span class="s1">&#39;activeSg&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
<span class="c1"># list of sgRNA which are true</span>
<span class="n">TruePositiveSGs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">indexPexActiveArray</span><span class="p">,</span> <span class="s1">&#39;sgID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">())</span>
</pre></div>
</div>
<p>Enrichment analysis was performed by comparing the activated simulated data (Q2) with the non-activated simulated data (Q1). Mapping of the read counts was done according to the unique Gene names, and the perturbation effects were calculated by dividing the log count data of Q2 by that of Q1.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">AIPyS</span> <span class="kn">import</span> <span class="n">mapSgRNA</span> 
<span class="c1"># filtered list of sgRNA data from the reference library for the sub library h4</span>
<span class="n">pathIn</span> <span class="o">=</span> <span class="sa">r</span><span class="s1">&#39;F:\HAB_2\PrinzScreen\AIPS_simulation\AIPS_simulation\data&#39;</span>
<span class="n">Q2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pathIn</span><span class="p">,</span><span class="s1">&#39;h4_1_45tpGuide_Q2.csv&#39;</span><span class="p">))</span>
<span class="n">Q1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pathIn</span><span class="p">,</span><span class="s1">&#39;h4_1_45tpGuide_Q1.csv&#39;</span><span class="p">))</span>
<span class="c1"># unique</span>

<span class="n">uniqueSgRNA</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">Q2</span><span class="o">.</span><span class="n">sgID</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="c1"># The mapping of unique sgRNA to the SIM was performed using the module mapsgrna</span>

<span class="n">dictDF</span> <span class="o">=</span> <span class="n">mapSgRNA</span><span class="p">(</span><span class="n">df1</span> <span class="o">=</span> <span class="n">Q1</span><span class="p">,</span> <span class="n">df2</span> <span class="o">=</span> <span class="n">Q2</span><span class="p">)</span>

<span class="c1"># Make dictionary mapping reads to unique sgRNA</span>

<span class="n">df_dict</span> <span class="o">=</span> <span class="n">dictDF</span><span class="o">.</span><span class="n">mapping</span><span class="p">(</span><span class="n">uniqueSgRNA</span> <span class="o">=</span> <span class="n">uniqueSgRNA</span><span class="p">)</span>

<span class="c1"># data frame contain logFC and z-score logFC</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">dictDF</span><span class="o">.</span><span class="n">dataFrameFinal</span><span class="p">(</span><span class="n">df_dict</span><span class="p">)</span>
</pre></div>
</div>
<center><b><u>Simulation</u></b></center>
<p><img alt="png" src="_images/output_7_0_0.png" /></p>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Gil Kanfer, PhD.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>